blocks_reexecutor/blocks_reexecutor.go

```diff
@@ -84,25 +83,39 @@ func New(c *Config, blockchain *core.BlockChain, fatalErrChan chan error) *Block
 		start = chainStart
 		end = chainEnd
 	}
-	if start < chainStart {
-		log.Warn("state reexecutor's start block number is lower than genesis, resetting to genesis")
+	if start < chainStart || start > chainEnd {
+		log.Warn("invalid state reexecutor's start block number, resetting to genesis", "start", start, "genesis", chainStart)
 		start = chainStart
 	}
-	if end > chainEnd {
-		log.Warn("state reexecutor's end block number is greater than latest, resetting to latest")
+	if end > chainEnd || end < chainStart {
+		log.Warn("invalid state reexecutor's end block number, resetting to latest", "end", end, "latest", chainEnd)
 		end = chainEnd
 	}
 	if c.Mode == "random" && end != start {
-		if c.BlocksPerThread > end-start {
-			c.BlocksPerThread = end - start
+		// Reexecute a range of 10000 or (non-zero) c.BlocksPerThread number of blocks between start to end picked randomly
+		rng := uint64(10000)
+		if c.BlocksPerThread != 0 {
+			rng = c.BlocksPerThread
+		}
+		if rng > end-start {
+			rng = end - start
 		}
-		start += uint64(rand.Intn(int(end - start - c.BlocksPerThread + 1)))
-		end = start + c.BlocksPerThread
+		start += uint64(rand.Intn(int(end - start - rng + 1)))
+		end = start + rng
 	}
-	// inclusive of block reexecution [start, end]
-	if start > 0 {
+	// Inclusive of block reexecution [start, end]
+	// Do not reexecute genesis block i,e chainStart
+	if start > 0 && start != chainStart {
 		start--
 	}
+	// Divide work equally among available threads
+	if c.BlocksPerThread == 0 {
+		c.BlocksPerThread = 10000
+		work := (end - start) / uint64(c.Room)
+		if work > 0 {
+			c.BlocksPerThread = work
```

Comment:
I am not sure if modifying config is the best way here, especially that we don't make a copy

probably cleaner way is to add extra field to BlocksReExecutor struct that will be set to `work` or `c.BlocksPerThread`

Reply:
or we may just make a copy of the config (?)